src/native/corehost/bundle/extractor.cpp

```diff
@@ -103,11 +113,68 @@ void extractor_t::extract(const file_entry_t &entry, reader_t &reader)
 {
     FILE* file = create_extraction_file(entry.relative_path());
     reader.set_offset(entry.offset());
-    int64_t size = entry.size();
+    size_t size = entry.size();
     size_t cast_size = to_size_t_dbgchecked(size);
-    if (fwrite(reader, 1, cast_size, file) != cast_size)
+    size_t extracted_size = 0;
+
+    if (entry.compressedSize() != 0)
+    {
+#if defined(NATIVE_LIBS_EMBEDDED)
+        PAL_ZStream zStream;
+        zStream.nextIn = (uint8_t*)(const void*)reader;
+        zStream.availIn = entry.compressedSize();
+
+        const int Deflate_DefaultWindowBits = -15; // Legal values are 8..15 and -8..-15. 15 is the window size,
+                                                   // negative val causes deflate to produce raw deflate data (no zlib header).
+
+        int ret = CompressionNative_InflateInit2_(&zStream, Deflate_DefaultWindowBits);
+        if (ret != PAL_Z_OK)
+        {
+            trace::error(_X("Failure initializing zLib stream."));
+            throw StatusCode::BundleExtractionIOError;
+        }
+
+        const int bufSize = 4096;
+        uint8_t* buf = (uint8_t*)alloca(bufSize);
+
+        do
+        {
+            zStream.nextOut = buf;
+            zStream.availOut = bufSize;
+
+            ret = CompressionNative_Inflate(&zStream, PAL_Z_NOFLUSH);
+            if (ret < 0)
+            {
+                CompressionNative_InflateEnd(&zStream);
+                trace::error(_X("Failure inflating zLib stream. %s"), zStream.msg);
+                throw StatusCode::BundleExtractionIOError;
+            }
+
+            int produced = bufSize - zStream.availOut;
+            if (fwrite(buf, 1, produced, file) != (size_t)produced)
+            {
+                CompressionNative_InflateEnd(&zStream);
+                trace::error(_X("I/O failure when writing decompressed file."));
+                throw StatusCode::BundleExtractionIOError;
+            }
+
+            extracted_size += produced;
+        } while (zStream.availOut == 0);
+
+        CompressionNative_InflateEnd(&zStream);
+#else
+        trace::error(_X("Compressed file in a standalone host scenario?"));
```

Comment:
Could this print any info. on what failed? I think it'd be a little bit difficult for users to understand where things went wrong from this message.

Reply:
Users are not supposed to see this message. It means that the bundle is corrupted since we do not support and should never produce such combination (not singlefile apphost + compression).

I think, I can make the error a bit more clear, but it is mostly for us - in case we regress something.